include/picongpu/fields/background/templates/twtsfast/BField.tpp

```diff
@@ -0,0 +1,738 @@
+                {
+                    for(uint32_t i = 0; i < simDim; ++i)
+                    {
+                        pos[k][i] = bFieldPositions_SI[k][i];
+                    }
+                }
+
+                /* An example of intra-cell position offsets is the staggered Yee-grid.
+                 *
+                 * Calculate By-component with the intra-cell offset of a By-field
+                 */
+                float_64 const By_By = calcTWTSBy(pos[1], time);
+                /* Calculate Bz-component the the intra-cell offset of a By-field */
+                float_64 const Bz_By = calcTWTSBz_Ex(pos[1], time);
+                /* Calculate By-component the the intra-cell offset of a Bz-field */
+                float_64 const By_Bz = calcTWTSBy(pos[2], time);
+                /* Calculate Bz-component the the intra-cell offset of a Bz-field */
+                float_64 const Bz_Bz = calcTWTSBz_Ex(pos[2], time);
+                /* Since we rotated all position vectors before calling calcTWTSBy and calcTWTSBz_Ex,
+                 * we need to back-rotate the resulting B-field vector.
+                 *
+                 * RotationMatrix[-(PI/2+phi)].(By,Bz) for rotating back the field vectors.
+                 */
+                float_64 const By_rot = -math::sin(+phi) * By_By + math::cos(+phi) * Bz_By;
+                float_64 const Bz_rot = -math::cos(+phi) * By_Bz - math::sin(+phi) * Bz_Bz;
+
+                /* Finally, the B-field normalized to the peak amplitude. */
+                return float3_X(float_X(0.0), float_X(By_rot), float_X(Bz_rot));
+            }
+
+            template<>
+            HDINLINE float3_X BField::getTWTSBfield_Normalized_Ey<DIM3>(
+                pmacc::math::Vector<floatD_64, detail::numComponents> const& bFieldPositions_SI,
+                float_64 const time) const
+            {
+                typedef pmacc::math::Vector<float3_64, detail::numComponents> PosVecVec;
+                PosVecVec pos(PosVecVec::create(float3_64::create(0.0)));
+
+                for(uint32_t k = 0; k < detail::numComponents; ++k)
+                {
+                    for(uint32_t i = 0; i < simDim; ++i)
+                    {
+                        pos[k][i] = bFieldPositions_SI[k][i];
+                    }
+                }
+
+                /* Calculate Bz-component with the intra-cell offset of a By-field */
+                float_64 const Bz_By = calcTWTSBz_Ey(pos[1], time);
+                /* Calculate Bz-component with the intra-cell offset of a Bz-field */
+                float_64 const Bz_Bz = calcTWTSBz_Ey(pos[2], time);
+                /* Since we rotated all position vectors before calling calcTWTSBz_Ey,
+                 * we need to back-rotate the resulting B-field vector.
+                 *
+                 * RotationMatrix[-(PI/2+phi)].(By,Bz) for rotating back the field-vectors.
+                 */
+                float_64 const By_rot = +math::cos(+phi) * Bz_By;
```

Comment:
if you write: `float_X const By_rot = +math::cos(+phi) * Bz_By;` you can avoid the the casts in line `155` which makes the code better readable.